File: leetcode/Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > zigzagLevelOrder(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		vector<vector<int> > ans;
		traverse(root, 0, ans);
		for (int i = 1; i < ans.size(); i += 2)
		{
			reverse(ans[i].begin(), ans[i].end());
		}
		return ans;
    }
    void traverse(const TreeNode *root, int level, vector<vector<int> > &ans)
    {
		if (NULL == root)
		{
			return;
		}
		while (ans.size() <= level)
		{
			ans.push_back(vector<int>());
		}
		ans[level].push_back(root->val);
		traverse(root->left, level + 1, ans);
		traverse(root->right, level + 1, ans);
	}
};

File: answers/leetcode/Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """

        level_list = []
        self.dfs(0, root, level_list)
        return level_list

    def dfs(self, level, root, level_list):
        if None == root:
            return
        while len(level_list) <= level:
            level_list.append([])
        if level % 2:
            level_list[level].insert(0, root.val)
        else:
            level_list[level].append(root.val)
        self.dfs(level + 1, root.left, level_list)
        self.dfs(level + 1, root.right, level_list)
